### ISSUE TITLE
Letter avatars and Gravatars not working

### ISSUE THREAD

**Issue by bkraul** [bitnami/bitnami-docker-discourse#57]
**Description**

Did a fresh docker-compose install of bitnami/discourse (See my docker-compose below). The only difference is that I am using alternate redis and postgres which to the best of my knowledge are operating as expected. Sidekiq reports it's working. But I am not seeing letter avatars or Gravatars show up. This is what the log is telling me:

```
Failed to process hijacked response correctly : /opt/bitnami/discourse/lib/letter_avatar.rb:85:in `generate_fullsize': convert-im6.q16: unable to read font `Helvetica' @ error/annotate.c/RenderFreetyp
----
/opt/bitnami/discourse/lib/discourse.rb:28:in `execute_command'
/opt/bitnami/discourse/lib/letter_avatar.rb:85:in `generate_fullsize'
/opt/bitnami/discourse/lib/letter_avatar.rb:46:in `block in generate'
/opt/bitnami/discourse/lib/distributed_mutex.rb:34:in `synchronize'
/opt/bitnami/discourse/lib/distributed_mutex.rb:5:in `synchronize'
/opt/bitnami/discourse/lib/letter_avatar.rb:34:in `generate'
/opt/bitnami/discourse/app/controllers/user_avatars_controller.rb:51:in `block in show_proxy_letter'
/opt/bitnami/discourse/lib/hijack.rb:56:in `instance_eval'
/opt/bitnami/discourse/lib/hijack.rb:56:in `block in hijack'
/opt/bitnami/discourse/lib/scheduler/defer.rb:70:in `block in do_work'
/opt/bitnami/discourse/vendor/bundle/ruby/2.4.0/gems/rails_multisite-2.0.4/lib/rails_multisite/connection_management.rb:63:in `with_connection'
/opt/bitnami/discourse/lib/scheduler/defer.rb:68:in `do_work'
/opt/bitnami/discourse/lib/scheduler/defer.rb:59:in `block (2 levels) in start_thread'
```

**My docker-compose.yml**
```yaml
version: '3.2'
services:
  postgresql:
    image: postgres:alpine
    restart: always
    volumes:
      - /root/docker-persist/discourse/postgresql/pgdata:/var/lib/postgresql/data/pgdata
    environment:
      - PGDATA=/var/lib/postgresql/data/pgdata
      - POSTGRES_INITDB_ARGS=--data-checksums --locale=en_US
    networks:
      - default
  redis:
    image: redis:alpine
    restart: always
    volumes:
      - discourse_redis:/data
    sysctls:
      - net.core.somaxconn=511
    networks:
      - default
  discourse:
    image: bitnami/discourse:latest
    restart: always
    volumes:
      - /root/docker-persist/discourse:/bitnami
    environment:
      - VIRTUAL_HOST=discourse.belmankraul.com
      - VIRTUAL_PORT=3000
      - DISCOURSE_USERNAME=admin
      - DISCOURSE_PASSWORD=******
      - DISCOURSE_EMAIL=[my discourse email]
      - DISCOURSE_SITENAME=bkraul
      - DISCOURSE_HOSTNAME=discourse.belmankraul.com
      - POSTGRESQL_CLIENT_CREATE_DATABASE_NAME=discourse
      - POSTGRESQL_CLIENT_CREATE_DATABASE_USERNAME=discourse
      - POSTGRESQL_CLIENT_CREATE_DATABASE_PASSWORD=******
      - DISCOURSE_POSTGRESQL_NAME=discourse
      - DISCOURSE_POSTGRESQL_USERNAME=discourse
      - DISCOURSE_POSTGRESQL_PASSWORD==******
      - SMTP_HOST=[my mail host]
      - SMTP_PORT=587
      - SMTP_USER=[my mail user]
      - SMTP_PASSWORD=******
    depends_on:
      - postgresql
      - redis
    networks:
      - reverse-proxy
      - default
  sidekiq:
    image: bitnami/discourse:latest
    restart: always
    volumes:
      - /root/docker-persist/discourse:/bitnami
    environment:
      - DISCOURSE_HOST=discourse
      - DISCOURSE_PORT=3000
      - DISCOURSE_POSTGRESQL_NAME=discourse
      - DISCOURSE_POSTGRESQL_USERNAME=discourse
      - DISCOURSE_POSTGRESQL_PASSWORD==******
      - DISCOURSE_SITENAME=bkraul
      - DISCOURSE_HOSTNAME=discourse.belmankraul.com
      - SMTP_HOST=[my mail host]
      - SMTP_PORT=587
      - SMTP_USER=[my mail user]
      - SMTP_PASSWORD==******
    command: 'nami start --foreground discourse-sidekiq'
    depends_on:
      - discourse
    networks:
      - default
volumes:
  discourse_redis:
networks:
  reverse-proxy:
    external:
      name: reverse-proxy
```

**Steps to reproduce the issue:**

1. Install bitnami/discourse with compose as shown
2. Log in with intial account

**Describe the results you received:**

There are no letter avatars for users, and creating a user that has an associated Gravatar account also does not show it:
![image](https://user-images.githubusercontent.com/14616851/47045967-b6903a00-d159-11e8-87e2-80c0934709e9.png)

**Describe the results you expected:**
Expected letter avatars and gravatars to work out of the box.

**Version**

- Output of `docker version`:

```
Client:
 Version:           18.06.1-ce
 API version:       1.38
 Go version:        go1.10.3
 Git commit:        e68fc7a
 Built:             Tue Aug 21 17:24:51 2018
 OS/Arch:           linux/amd64
 Experimental:      false

Server:
 Engine:
  Version:          18.06.1-ce
  API version:      1.38 (minimum version 1.12)
  Go version:       go1.10.3
  Git commit:       e68fc7a
  Built:            Tue Aug 21 17:23:15 2018
  OS/Arch:          linux/amd64
  Experimental:     false
```

- Output of `docker info`:

```
Containers: 30
 Running: 30
 Paused: 0
 Stopped: 0
Images: 72
Server Version: 18.06.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 468a545b9edcd5932818eb9de8e72413e616e86e
runc version: 69663f0bd4b60df09991c08812a60108003fa340
init version: fec3683
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.15.0-36-generic
Operating System: Ubuntu 18.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 7.79GiB
Name: docker
ID: X777:P7YR:LILB:TJV4:YWAV:MBHI:MPIA:GSYV:MUJ2:5YLD:PJOU:K2MP
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Username: bkraul
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

- Output of `docker-compose version` (if applicable):

```
docker-compose version 1.22.0, build f46880f
docker-py version: 3.4.1
CPython version: 2.7.15rc1
OpenSSL version: OpenSSL 1.1.0g  2 Nov 2017
```

**Additional environment details (AWS, VirtualBox, Docker for MAC, physical, etc.):**
VMWare ESXi VM:
* 8GB RAM
* 4 VCPUs
* 60GB HD

**Comment by bkraul** [comment]
I went ahead and opted to try to get the official distro working. Answering this might help someone else if it is eventually answered, but I have moved on.

**Comment by juan131** [comment]
Hi @bkraul 

It looks like a issue with the fonts:

```
Failed to process hijacked response correctly : /opt/bitnami/discourse/lib/letter_avatar.rb:85:in `generate_fullsize': convert-im6.q16: unable to read font `Helvetica' @ error/annotate.c/RenderFreetyp
```

**convert-im6.q16** is part of ImageMagick an it seems it's unable to read the fonts for some reason. I'm going to open a task to investigate it. Thanks so much for reporting it.

**Comment by mebibou** [comment]
I have the same issue with fresh install from a few days ago, gravatars are not showing at all although for the same email address they show on other websites

**Comment by opensorceror** [comment]
+1, I have this issue too.

**Comment by marlinhares** [comment]
On this tag 2.2.5-debian-9-r18, Gravatar worked for me

**Comment by juan131** [comment]
I wasn't able to reproduce the issue either with the latest version.

@opensorceror @bkraul could you please confirm you're not facing issues with that version either?

**Comment by bkraul** [comment]
As mentioned before, I no longer use the bitnami image and instead use the discourse stand-alone docker distribution.

**Comment by prbsparx** [comment]
We're running into the same issue described here. 
We're using the bitnami discourse:2.1.8 docker image in kubernetes due to the version of postgres we have available in our managed database service.

The main common item I see is "Failed to process hijacked response correctly : /opt/bitnami/discourse/lib/letter_avatar.rb:85:in `generate_fullsize': convert-im6.q16: unable to read font `Helvetica' @ error/annotate.c/RenderFreetyp"

`/opt/bitnami/discourse-sidekiq/app/assets/fonts` doesn't have a Helvetica font. Not sure where ImageMagick pulls it from...

**Comment by juan131** [comment]
Hi @prbsparx 

That's a pretty old image @prbsparx  current available image is **2.3.9**). Are you able to reproduce the issue in current versions of the image?

**Comment by prbsparx** [comment]
Yes, I will try that. I did look through the dockerfile to see if you install Helvetica, and it looks like you don't. The font is hardcoded into Discourse's app.

My kubernetes hosts aren't MacOS... and the docker image is Debian, I figured I'd point out that the Discourse dev instructions for MacOS mention you have to install Helvetica... https://github.com/discourse/discourse/blob/a4f5f5d0bb900539d0ea6a045b25d2050b9fe122/docs/DEVELOPMENT-OSX-NATIVE.md

I'll try to reproduce with 2.3.9 but it'll take a little bit as I have to update a few things in our Helm chart.

**Comment by juan131** [comment]
Hi @prbsparx 

It is very likely that we need to build the fonts for Helvetica (even if the container is Debian).

Please keep us updated about what you find out using the latest container 🤞

**Comment by prbsparx** [comment]
We're working on testing with 2.3.9... hope to have an answer for you later today.

**Comment by prbsparx** [comment]
Well bugger... letter avatars are working in 2.3.9... so why is it failing in 2.1.8...

If we have to run PostgreSQL 9.6, what version of bitnami discourse would you recommend we run?

**Comment by juan131** [comment]
Hi @prbsparx 

At this moment we test this Discourse version with PostgreSQL 11 (`bitnami/postgresql:11`) but we also maintain the branches **9.6**, **10** and **12** for PostgreSQL.

According to [Discourse Install Guide](https://github.com/discourse/discourse/blob/master/docs/INSTALL.md) you should be using PostgreSQL 10+

**Comment by prbsparx** [comment]
@juan131 - The problem I'm running into is that the database service we use doesn't have 10, 11, or 12 yet. We were hoping to use the database service that we own...

When we tested Discourse 2.3.8 with Postgres 9.6 we couldn't import the database because the postgres client in the discourse image is postgres 11. 
We verified that 9.6 isn't supported by Discourse so we decided to try an older discourse image. And that's when we ran into avatar issues.

**Comment by juan131** [comment]
Hi @prbsparx 

An alternative would be creating your own Discourse image using the "old" Discourse version. To do so:

- Clone the repo and go to latest commit we added for `2.1.8` image. Then go to the `2/debian-9` folder:

```console
$ git clone https://github.com/bitnami/bitnami-docker-discourse.git
$ git checkout 8a523f4b761946c56b697705e56ec040974c11cb
$ cd bitnami-docker-discourse/2/debian-9
```

- Build your custom image:

```console
$ docker build . -t my-custom-discourse:2.1.8
```

It's very likely that it install the latest version of imagemagick and the problem should be fixed.

**Comment by marcosbc** [comment]
Hi, FYI we just released a new container version, [2.7.0-debian-10-r4](https://github.com/bitnami/bitnami-docker-discourse#270-debian-10-r4). The initialization logic has received a major refactor, and its scripts are now located in the [`rootfs/` directory](https://github.com/bitnami/bitnami-docker-discourse/blob/master/2/debian-10/rootfs/).

We recommend to check if this issue is still happening in this version. If it is, make sure that you are using port 80/443 for HTTP/HTTPS respectively, because unfortunately Discourse isn't officially supported in any other port and we have seen things like issues in letter avatars happening if the ports are different.
